### DISCUSSION TITLE
connection lost when use eventlet Timeout

### DISCUSSION THREAD

**Discussion by Archer-Code** [sqlalchemy/sqlalchemy#8966] [Usage Questions]
### Describe the bug

It seems that the implementation of QueuePool has the risk of connection leakage.  When multi coroutine call QueuePool._do_return_conn, them would be block on by on. On the other hand, when obtaining a connection, it attempts to notify the object waiting to release the connection.  When the Timeout event of the eventlet is combined, an exception may occur in the coroutine that obtains the connection successfully and is not wrapped by _ConnectionFairy.  

### To Reproduce

I hope I'm mistaken, but let's see if there's a real risk. I haven't written a suitable example to verify this.

### Error

```
# Copy complete stack trace and error message here, including SQL log output if applicable.

File "/usr/lib64/python3.7/site-packages/sqlalchemy/orm/query.py", line 3228, in first
    ret = list(self[0:1])
  File "/usr/lib64/python3.7/site-packages/sqlalchemy/orm/query.py", line 3018, in __getitem__
    return list(res)
  File "/usr/lib64/python3.7/site-packages/sqlalchemy/orm/query.py", line 3330, in __iter__
    return self._execute_and_instances(context)
  File "/usr/lib64/python3.7/site-packages/sqlalchemy/orm/query.py", line 3352, in _execute_and_instances
    querycontext, self._connection_from_session, close_with_result=True
  File "/usr/lib64/python3.7/site-packages/sqlalchemy/orm/query.py", line 3367, in _get_bind_args
    mapper=self._bind_mapper(), clause=querycontext.statement, **kw
  File "/usr/lib64/python3.7/site-packages/sqlalchemy/orm/query.py", line 3345, in _connection_from_session
    conn = self.session.connection(**kw)
  File "/usr/lib64/python3.7/site-packages/sqlalchemy/orm/session.py", line 1124, in connection
    execution_options=execution_options,
  File "/usr/lib64/python3.7/site-packages/sqlalchemy/orm/session.py", line 1133, in _connection_for_bind
    conn = engine._contextual_connect(**kw)
  File "/usr/lib64/python3.7/site-packages/sqlalchemy/engine/base.py", line 2229, in _contextual_connect
    self._wrap_pool_connect(self.pool.connect, None),
  File "/usr/lib64/python3.7/site-packages/sqlalchemy/engine/base.py", line 2265, in _wrap_pool_connect
    return fn()
  File "/usr/lib64/python3.7/site-packages/sqlalchemy/pool/base.py", line 365, in connect
    return _ConnectionFairy._checkout(self)
  File "/usr/lib64/python3.7/site-packages/sqlalchemy/pool/base.py", line 766, in _checkout
    fairy = _ConnectionRecord.checkout(pool)
  File "/usr/lib64/python3.7/site-packages/sqlalchemy/pool/base.py", line 496, in checkout
    rec = pool._do_get()
  File "/usr/lib64/python3.7/site-packages/sqlalchemy/pool/impl.py", line 140, in _do_get
    code="3o7r",
sqlalchemy.exc.TimeoutError: QueuePool limit of size 10 overflow 10 reached, connection timed out, timeout 30 (Background on this error at: http://sqlalche.me/e/3o7r)
: sqlalchemy.exc.TimeoutError: QueuePool limit of size 10 overflow 10 reached, connection timed out, timeout 30 (Background on this error at: http://sqlalche.me/e/3o7r)
```
### Versions

- OS: Linux MNG-2 4.18.0-147.5.2.10.h913.eulerosv2r10.x86_64 #1 SMP Thu May 19 08:23:57 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
- Python: 3.7.9
- SQLAlchemy: 1.3.6
- Database: postgresql
- DBAPI (eg: psycopg, cx_oracle, mysqlclient):

**Comment by CaselIT** [comment]
Hi,

Moved to discussion since there is no information for this to be an issue.
I don't understand what you mean by
> When multi coroutine call QueuePool._do_return_conn, them would be block on by on

sqlalchemy 1.3.x does not support asyncio, so no coroutine would call that method

**Reply by Archer-Code** [reply]
use eventlet , this can create multi coroutine and support timeout when a coroutine running over time.

**Reply by CaselIT** [reply]
I've usually heard them called "green thread", not coroutine.

Still I'm not sure what you are asking. how would a connection leak?

**Reply by Archer-Code** [reply]
Please take a look at my reply below.

**Comment by zzzeek** [comment]
just to note, the stack trace shows SQLAlchemy's own TimeoutError, and is not an eventlet-related error.

When using eventlet or gevent with the plain QueuePool, we are using eventlet/greenlet monkeypatching.   These functions provide the hypothetical guarantee that any code that works under traditional threading, including locking and concurrency scenarios, will behave the same way under greenlet monkeypatching.     So any race condition you could come up with under eventlet would be theoretically observable with plain threads, however the practical difference is that some race conditions that are not generally reproducible under threads are in fact reproducible under greenlet monkeypatching, as the scheduling behavior is totally different.   

This is to say there should not be any theoretical race that requires greenlet monkeypatching to be in place unless it involves specific exceptions local to greenlet.     There was in fact one such issue, which was fixed in SQLAlchemy 1.1 which is #3803.

re: connection "leakage", this is usually not possible in practice even if QueuePool completely failed at its task, subject to the details of the DBAPI driver in use, as most DBAPIs will release underlying sockets to the server when the connection is garbage collected.    

I've tried to follow the specific scenario you're laying out but I'm not following it.  For QueuePool, both `_do_get()` and `_do_return_conn()` are blocking on the same mutex which is the `Queue.not_empty` mutex.  This is based on the code for `Queue` from Python standard library.    Eventlet has its own `Queue` class that is not monkeypatched here however evenlet would monkeypatch `threading.Condition` to work under non-blocking IO.

Overall, I've worked with SQLAlchemy under eventlet in Openstack for about eight years and #3803 was the only eventlet-related race condition we had to deal with.

**Reply by Archer-Code** [reply]
I mean coroutine  is a green thread.  I'm going to describe my point in as much detail as possible, partly by machine translation because I'm not good at it.  As you said  the stack trace is SQLAlchemy's own TimeoutError .  This is caused by a leak in the QueuePool. By connection leak, I mean that the connection is not put back into the connection pool.  When multiple green threads release connections to the queue, it will notify and switch to other green threads waiting to get the connection. At the same time, the green thread after getting the connection from the queue will also notify and switch to other waiting for release connection to the green threads. If some of these green threads is set to timeout by eventlet, the green threads that get the connection from queue may be unexpectedlly thrown by the eventlet after the notification ends. At this time. the connection is not wrapped by _ConnectionFairy and cannot be automatically recoverd.

**Reply by zzzeek** [reply]
>  If some of these green threads is set to timeout by eventlet, the green threads that get the connection from queue may be unexpectedlly thrown by the eventlet after the notification ends. 

I dont understand how?   the timeout affects the green thread in which it was invoked.  other green threads are not interrupted.  it's also not clear when you refer to "timeout by eventlet" if you mean eventlet's actual timeout, or the Timeout that SQLAlchemy throws, or some combination of the two.   

it's also not clear what your real world problem is and how it manifests.  Are you seeing too many connections being opened?  or is the pool simply hanging once it has "leaked" all connections?  What kind of application ?  how is it run?  are there conditions that are producing a large frequency of eventlet timeouts?  how often does the problem happen?

addtionally, why SQLAlchemy 1.3.6?  at the very least you should be upgrading to the most recent 1.3.x version which is 1.3.24, and additionally if we were to identify and fix any issue here, it would only be in the 1.4.x series, so upgrading to the latest 1.4.x series is also something you should be seeking to do. 

in any case, you would need to construct a self-contained demonstration case that illustrates the failure.   there's no other way we could move forward.   See https://gist.github.com/zzzeek/947337f44d70fac537be linked from #3258 for an example of such a test case, as well as all the attachments at the top of #3803.

**Reply by zzzeek** [reply]
"At the same time, the green thread after getting the connection from the queue will also notify and switch to other waiting for release connection to the green threads."   - dont understand that either - if a greenthread acquires the connection, it has it, it's atomic from the queue.   it doesn't yield until IO occurs on the connection which is afterwards. IO is not involved in the actual "put" / "get" that's synchronized on the queue.

**Reply by Archer-Code** [reply]
###
LOG： session status:Pool size: 5  Connections in pool: 5 Current Overflow: 1 Current Checked out connections: 1
lsof -i:5432 -anPp 50026
COMMAND     PID      USER   FD   TYPE    DEVICE SIZE/OFF NODE NAME
/usr/bin/ 50026 openstack   16u  IPv4 816283722      0t0  TCP 172.28.0.2:52680->172.28.9.30:5432 (ESTABLISHED)
/usr/bin/ 50026 openstack   17u  IPv4 816324636      0t0  TCP 172.28.0.2:53008->172.28.9.30:5432 (ESTABLISHED)
/usr/bin/ 50026 openstack   18u  IPv4 816275678      0t0  TCP 172.28.0.2:52670->172.28.9.30:5432 (ESTABLISHED)
/usr/bin/ 50026 openstack   19u  IPv4 816295970      0t0  TCP 172.28.0.2:52804->172.28.9.30:5432 (ESTABLISHED)
/usr/bin/ 50026 openstack   20u  IPv4 816266744      0t0  TCP 172.28.0.2:52666->172.28.9.30:5432 (ESTABLISHED)
#### 
Current Overflow is always 1, expect 0

**Reply by zzzeek** [reply]
there's a similar sounding issue from many years ago at https://github.com/urllib3/urllib3/issues/805.   but the test script is very simple and does not produce any problem if a SQLAlhcemy engine is substituted in there.      try working with this

```py
from gevent import monkey, spawn, sleep, Timeout, joinall
monkey.patch_all()

import sys
from itertools import product
from sqlalchemy import create_engine, text

def timed_read(engine, i, timeout, release=None, preload_content=True):
    try:
        with Timeout(timeout):
            conn = engine.connect()

            try:
                print ("%s opened" % i)
                result = conn.execute(text("select 1"))
                data = result.fetchall()
            finally:
                conn.close()
    except Timeout:
        print ("%s timed out" % i)
    else:
        pass


def main(timeout=10, release=None, preload=True, poolsize=2):
    print('-' * 10 + "  timeout=%s preload=%s release=%s " % (timeout, preload, release) + '-' * 10)
    engine = create_engine("mysql+pymysql://scott:tiger@localhost/test", pool_size=poolsize, pool_timeout=3)

    for j in range(200):
        group = []
        for i in range(3):
            group.append(spawn(timed_read, engine, i, timeout, release=release, preload_content=preload))
        joinall(group)


for timeout, preload in product([10, 4, 0.1], [True, False]):
    main(timeout=timeout, preload=preload)
```

**Reply by zzzeek** [reply]
since you are working with Openstack, what openstack app?  can you refer to a launchpad issue ?

**Reply by zzzeek** [reply]
funny thing the above program has a problem in main but not 1.4.  now *that* is weird!

**Reply by zzzeek** [reply]
oh, nevermind, that's only if I take the .close() out.   that's due to future engine mode which is more strict about closing connections (more specifically it's doing autobegin which is preventing conn from being immediately GCed if close is not called.  calling begin() explicitly produces the same thing in 1.3.6, 1.4).

**Reply by Archer-Code** [reply]
### QueuePool
    def _do_get(self):
        use_overflow = self._max_overflow > -1

        try:
            wait = use_overflow and self._overflow >= self._max_overflow
            return self._pool.get(wait, self._timeout)
        ......
###  util.Queue
     get()
         ...
         item = self._get()
         self.not_full.notify()
         return item
just like above， the eventlet timeout would be call greenlet.throw after not_full.notify end. then the item lost.  I still can't give a good example at the moment.

**Reply by CaselIT** [reply]
> the eventlet timeout would be call greenlet.throw after not_full.notify end. then the item lost. 

I may be wrong, but I don't think notify does any IO, so I don't see how greenlet can timeout after it

**Reply by Archer-Code** [reply]
###  Condition
    def __init__(self, lock=None):
        if lock is None:
            lock = RLock()
        self._lock = lock
        # Export the lock's acquire() and release() methods
        self.acquire = lock.acquire
        self.release = lock.release

### Semaphore
    def acquire(self, blocking=True, timeout=None):
        ...
        # If someone else is already in this wait loop, give them
        # a chance to get out.
        while True:
            hubs.get_hub().switch()
            if self.counter > 0:
                break
    def release(self, blocking=True):
        self.counter += 1
        if self._waiters:
            hubs.get_hub().schedule_call_global(0, self._do_acquire)
        return True

### Timeout
    def start(self):
        """Schedule the timeout.  This is called on construction, so
        it should not be called explicitly, unless the timer has been
        canceled."""
        ...
        else:  # regular timeout with user-provided exception
            self.timer = get_hub().schedule_call_global(
                self.seconds, greenlet.getcurrent().throw, self.exception)
        return self

eventlet make threading.Condition green,  eventlet.semaphore.Semaphore.acquire == Queue.not_empty.acquire == Queue.not_full.acquire , so no any IO in notify,  there is a switchover of green threads.

**Reply by CaselIT** [reply]
sure, but notify does not call acquire. It calls `release()` that does `hubs.get_hub().schedule_call_global(0, self._do_acquire)` that from what I can understand does not do any switch.

**Reply by Archer-Code** [reply]
may be this can help you understand what i mean
```
query err=<class '__main__.DBTimeout'>, stack=Traceback (most recent call last):
  File "test_event_timeout.py", line 68, in query_with_timeout
    _query(maker, index=9999)
  File "test_event_timeout.py", line 47, in _query
    result = session.execute(text("select count(*) from StandardAttributes where resource_type='test'"))
  File "/usr/lib64/python3.7/site-packages/sqlalchemy/orm/session.py", line 1268, in execute
    return self._connection_for_bind(bind, close_with_result=True).execute(
  File "/usr/lib64/python3.7/site-packages/sqlalchemy/orm/session.py", line 1133, in _connection_for_bind
    conn = engine._contextual_connect(**kw)
  File "/usr/lib64/python3.7/site-packages/sqlalchemy/engine/base.py", line 2229, in _contextual_connect
    self._wrap_pool_connect(self.pool.connect, None),
  File "/usr/lib64/python3.7/site-packages/sqlalchemy/engine/base.py", line 2265, in _wrap_pool_connect
    return fn()
  File "/usr/lib64/python3.7/site-packages/sqlalchemy/pool/base.py", line 363, in connect
    return _ConnectionFairy._checkout(self)
  File "/usr/lib64/python3.7/site-packages/sqlalchemy/pool/base.py", line 760, in _checkout
    fairy = _ConnectionRecord.checkout(pool)
  File "/usr/lib64/python3.7/site-packages/sqlalchemy/pool/base.py", line 492, in checkout
    rec = pool._do_get()
  File "/usr/lib64/python3.7/site-packages/sqlalchemy/pool/impl.py", line 118, in _do_get
    return self._pool.get(wait, self._timeout)
  File "/usr/lib64/python3.7/site-packages/sqlalchemy/util/queue.py", line 162, in get
    self.not_empty.wait(remaining)
  File "/usr/lib64/python3.7/threading.py", line 300, in wait
    gotit = waiter.acquire(True, timeout)
  File "/usr/lib/python3.7/site-packages/eventlet/semaphore.py", line 107, in acquire
    hubs.get_hub().switch()
  File "/usr/lib/python3.7/site-packages/eventlet/hubs/hub.py", line 313, in switch
    return self.greenlet.switch()
DBTimeout
```

**Reply by Archer-Code** [reply]
### this is my exapmple ,  make a  delay after get connection from queue
```py
import logging
import time
import traceback

import eventlet

from sqlalchemy import create_engine
from sqlalchemy import text
from sqlalchemy.orm import Query
from sqlalchemy.orm import Session
from sqlalchemy.orm import sessionmaker


eventlet.monkey_patch()


LOG = logging.getLogger(__name__)
LOG.addHandler(logging.StreamHandler())
LOG.setLevel("INFO")


def get_url():
    cfg = {
        "user": "openstack",
        "ps": "FusionSphere123",
        "addr": "[fc00:ac1c::81b]",
        "port": 5432,
        "db": "neutron"
    }
    return "postgresql://%(user)s:%(ps)s@%(addr)s:%(port)s/%(db)s" % cfg


class DBTimeout(BaseException):
    def __init__(self):
        super(DBTimeout, self).__init__()


def waste_cpu_time(n):
    for i in range(n):
        i += 1


def _query(maker, index=1):
    session = maker()
    cost = time.monotonic()
    result = session.execute(text("select count(*) from StandardAttributes where resource_type='test'"))
    data = [x[0] for x in result][0]
    # for row in result:
    #     if not data:
    #         # cost cpu time 5s, not same in other machine.
    #         # waste_cpu_time(50000000)
    #         time.sleep(5)
    #     data.append(row)
    cost = time.monotonic() - cost
    session.close()
    LOG.info("index %s result size=%s cost %s", index, data, cost)
    time.sleep(0)


def query_with_timeout(maker, index=0, tm=None):
    while True:
        if not tm:
            _query(maker, index)
            continue
        try:
            with eventlet.Timeout(tm, DBTimeout):
                _query(maker, index=9999)
        except BaseException as e:
            LOG.info("query err=%s, stack=%s", type(e), traceback.format_exc())
        LOG.info("pool status=%s", maker().bind.pool.status())


def main():
    engine = create_engine(get_url(), pool_size=2, max_overflow=2, pool_timeout=2)
    maker = sessionmaker(bind=engine, class_=Session, autocommit=True,
                         expire_on_commit=False, query_cls=Query)

    for i in range(4):
        eventlet.spawn(query_with_timeout, maker, index=i)
    gp = eventlet.GreenPool(4)
    for i in range(4):
        gp.spawn(query_with_timeout, maker, index=999 + i, tm=2)
    gp.waitall()


if __name__ == '__main__':
    main()
```

## modify sqlachemy.util.queue 
     def get(self, block=True, timeout=None)
     ...
     item = self._get()
     # cost cpu time 1 second
     for i in range(10000000):
         i += 1
     self.not_full.notify()
     return item

**Reply by Archer-Code** [reply]
> sure, but notify does not call acquire. It calls `release()` that does `hubs.get_hub().schedule_call_global(0, self._do_acquire)` that from what I can understand does not do any switch.

```
def _do_acquire(self):
        if self._waiters and self.counter > 0:
            waiter = self._waiters.popleft()
            waiter.switch()
```

**Reply by CaselIT** [reply]
> > sure, but notify does not call acquire. It calls `release()` that does `hubs.get_hub().schedule_call_global(0, self._do_acquire)` that from what I can understand does not do any switch.
> 
> ```
> def _do_acquire(self):
>         if self._waiters and self.counter > 0:
>             waiter = self._waiters.popleft()
>             waiter.switch()
> ```

from what I understand `_do_acquire` is not executed in the caller greenlet, so the caller that executes `hubs.get_hub().schedule_call_global(0, self._do_acquire)` is not interrupted by the `_do_acquire`. Is that not the case?

**Reply by zzzeek** [reply]
I ran the program, adding print statements as:

```diff
diff --git a/lib/sqlalchemy/util/queue.py b/lib/sqlalchemy/util/queue.py
index 67c5219c72..508e08931c 100644
--- a/lib/sqlalchemy/util/queue.py
+++ b/lib/sqlalchemy/util/queue.py
@@ -157,6 +157,13 @@ class Queue:
                         raise Empty
                     self.not_empty.wait(remaining)
             item = self._get()
+
+            print("here's that delay")
+            # cost cpu time 1 second
+            for i in range(10000000):
+                i += 1
+            print("ok it's done")
+
             self.not_full.notify()
             return item
```

and then establishing a limit for how long the program runs, since the goal is, see if the pool was corrupted once nothing is running:

```py
def query_with_timeout(maker, index=0, tm=None):
    for i in range(5):
        if not tm:
            _query(maker, index)
            continue
        try:
            with eventlet.Timeout(tm, DBTimeout):
                _query(maker, index=9999)
        except BaseException as e:
            LOG.info("query err=%s, stack=%s", type(e))
        LOG.info("pool status=%s", maker().bind.pool.status())


def main():
    engine = create_engine(get_url(), pool_size=2, max_overflow=2, pool_timeout=2)
    maker = sessionmaker(
        bind=engine,
        class_=Session,
        autocommit=True,
        expire_on_commit=False,
        query_cls=Query,
    )

    for i in range(4):
        eventlet.spawn(query_with_timeout, maker, index=i)
    gp = eventlet.GreenPool(4)
    for i in range(4):
        gp.spawn(query_with_timeout, maker, index=999 + i, tm=2)
    gp.waitall()
    LOG.info("pool status=%s", maker().bind.pool.status())
    breakpoint()

```

so the print statements illustrate part of how eventlet works, is that **no other green threads can possibly do anything** during CPU bound operation.  meaning, the CPU wait that is added doesn't mean that other green threads are going to be able to go into an invalid state during that wait.  in the program's run we see many prints like:

```
here's that delay
ok it's done
index 3 result size=1 cost 0.8425949099473655
here's that delay
ok it's done

```

there's no instance where anything is printed in between "here's that delay" and "ok it's done" - it's not possible because there is only one real thread.  There's no IO happening in there (standard out doesn't count for python blocking IO).  

when the end of the program is reached, the greenlets have all had lots of problems due to all those timeouts and the very long wait time for a connection, sure, but the pool is stable:

```
pool status=Pool size: 2  Connections in pool: 2 Current Overflow: 0 Current Checked out connections: 0
--Return--
> /home/classic/dev/sqlalchemy/test3.py(92)main()->None
-> breakpoint()
```

so what we need to see is the thing you illustrated above, which is the conditoin you illustrated earlier:

```
LOG： session status:Pool size: 5 Connections in pool: 5 Current Overflow: 1 Current Checked out connections: 1
lsof -i:5432 -anPp 50026
COMMAND PID USER FD TYPE DEVICE SIZE/OFF NODE NAME
/usr/bin/ 50026 openstack 16u IPv4 816283722 0t0 TCP 172.28.0.2:52680->172.28.9.30:5432 (ESTABLISHED)
/usr/bin/ 50026 openstack 17u IPv4 816324636 0t0 TCP 172.28.0.2:53008->172.28.9.30:5432 (ESTABLISHED)
/usr/bin/ 50026 openstack 18u IPv4 816275678 0t0 TCP 172.28.0.2:52670->172.28.9.30:5432 (ESTABLISHED)
/usr/bin/ 50026 openstack 19u IPv4 816295970 0t0 TCP 172.28.0.2:52804->172.28.9.30:5432 (ESTABLISHED)
/usr/bin/ 50026 openstack 20u IPv4 816266744 0t0 TCP 172.28.0.2:52666->172.28.9.30:5432 (ESTABLISHED)
```

I need to see that, and most preferably, **please make an openstack issue for neutron on launchpad**, since I have now guessed you are using neutron, it is very likely this issue has been seen before and possibly even fixed, or if not, a real reproducer might be present.

**Reply by zzzeek** [reply]
here's that link:   https://bugs.launchpad.net/neutron

**Reply by zzzeek** [reply]
OK it may have reproduced after a few runs!   now I have something to look at.  thanks

**Comment by zzzeek** [comment]
OK we have the issue at #8974, where I spent lots of time printing things and making color coding to find things etc., which was mostly just "fun" I guess, because by doing that I noticed the problem is just the "failed checkout" catch, which we already have for this exact case, was as usual not catching BaseException, so this seems to fix:

```diff
diff --git a/lib/sqlalchemy/pool/base.py b/lib/sqlalchemy/pool/base.py
index a8234c5309..43e608c0fb 100644
--- a/lib/sqlalchemy/pool/base.py
+++ b/lib/sqlalchemy/pool/base.py
@@ -491,7 +491,7 @@ class _ConnectionRecord(object):
         rec = pool._do_get()
         try:
             dbapi_connection = rec.get_connection()
-        except Exception as err:
+        except BaseException as err:
             with util.safe_reraise():
                 rec._checkin_failed(err, _fairy_was_created=False)
         echo = pool._should_log_debug()

```

can you confirm this resolves for your end?  You will need to use SQLAlchemy 1.4 since we are not releasing new 1.3 versions anymore.

**Reply by Archer-Code** [reply]
yes, i am using neutron . In fact, the one point you've found may be one of the problems.  Maybe there's no problem with the implementation of the queue. I'm going to sort out the green thread switch, and it looks like the other green threads that are waiting to release the connection are not switched, so the eventlet call greenlet.throw is not triggered. Thank you very much. It took a lot of your time. I'm not sure if the change is working yet, it's hard to construct, and here's a log that clearly shows that there's a connection leak.
```
2022-12-13T09:44:20.437048+08:00 localhost neutron-server  INFO [pid:50026] [http_heart_beat][tid:22955654585872] =====session status:Pool size: 5  Connections in pool: 0 Current Overflow: 5 Current Checked out connections: 10
2022-12-13T09:45:31.262271+08:00 localhost neutron-server  ERROR [pid:50026] QueuePool limit of size 5 overflow 5 reached, connection timed out, timeout 30 (Background on this error at: http://sqlalche.me/e/3o7r): sqlalchemy.exc.TimeoutError: QueuePool limit of size 5 overflow 5 reached, connection timed out, timeout 30

## TCP connecton  0
MNG-1:~ # lsof -i:5432 -anPp 50026
MNG-1:~ # 

```

**Reply by zzzeek** [reply]
that's with the fix?    need a new test case then :)   the issue here was one that was already known, just wasn't catching all the exceptions.

**Reply by Archer-Code** [reply]
I'm not going to fix it yet. What you found led me to rework the code. I will continue to reproduce the problem, from the new log should see the problem, thank you for your help.

**Reply by Archer-Code** [reply]
I got the same error.  i will fix it. thanks !
```
 File "/usr/lib64/python3.7/site-packages/sqlalchemy/pool/base.py", line 627, in get_connection
    self.__connect()
  File "/usr/lib64/python3.7/site-packages/sqlalchemy/pool/base.py", line 645, in __connect
    connection = pool._invoke_creator(self)
  File "/usr/lib64/python3.7/site-packages/sqlalchemy/engine/strategies.py", line 114, in connect
    return dialect.connect(*cargs, **cparams)
  File "/usr/lib64/python3.7/site-packages/sqlalchemy/engine/default.py", line 453, in connect
    return self.dbapi.connect(*cargs, **cparams)
  File "/usr/lib64/python3.7/site-packages/psycopg2/__init__.py", line 126, in connect
    conn = _connect(dsn, connection_factory=connection_factory, **kwasync)
  File "/usr/lib/python3.7/site-packages/eventlet/support/psycopg2_patcher.py", line 52, in eventlet_wait_callback
    eventlet.hubs.trampoline(conn.fileno(), write=True)
  File "/usr/lib/python3.7/site-packages/eventlet/hubs/__init__.py", line 159, in trampoline
    return hub.switch()
  File "/usr/lib/python3.7/site-packages/eventlet/hubs/hub.py", line 313, in switch
    return self.greenlet.switch()
```

**Reply by CaselIT** [reply]
@Archer-Code you mean you reproduced the same error mentioned by mike without the fix or that you still have the same error with the fix proposed?

**Reply by Archer-Code** [reply]
reproduced the same error without fix.  I'm going to redefine an exception that sqlalchemy can catch. It's not verified yet.

**Reply by Archer-Code** [reply]
@zzzeek  There's another scenario that's wrong.
```
 File "/usr/lib64/python3.7/site-packages/sqlalchemy/orm/query.py", line 3352, in _execute_and_instances
    querycontext, self._connection_from_session, close_with_result=True
  File "/usr/lib64/python3.7/site-packages/sqlalchemy/orm/query.py", line 3367, in _get_bind_args
    mapper=self._bind_mapper(), clause=querycontext.statement, **kw
  File "/usr/lib64/python3.7/site-packages/sqlalchemy/orm/query.py", line 3345, in _connection_from_session
    conn = self.session.connection(**kw)
  File "/usr/lib64/python3.7/site-packages/sqlalchemy/orm/session.py", line 1124, in connection
    execution_options=execution_options,
  File "/usr/lib64/python3.7/site-packages/sqlalchemy/orm/session.py", line 1133, in _connection_for_bind
    conn = engine._contextual_connect(**kw)
  File "/usr/lib64/python3.7/site-packages/sqlalchemy/engine/base.py", line 2229, in _contextual_connect
    self._wrap_pool_connect(self.pool.connect, None),
  File "/usr/lib64/python3.7/site-packages/sqlalchemy/engine/base.py", line 2265, in _wrap_pool_connect
    return fn()
  File "/usr/lib64/python3.7/site-packages/sqlalchemy/pool/base.py", line 365, in connect
    return _ConnectionFairy._checkout(self)
  File "/usr/lib64/python3.7/site-packages/sqlalchemy/pool/base.py", line 766, in _checkout
    fairy = _ConnectionRecord.checkout(pool)
  File "/usr/lib64/python3.7/site-packages/sqlalchemy/pool/base.py", line 496, in checkout
    rec = pool._do_get()
  File "/usr/lib64/python3.7/site-packages/sqlalchemy/pool/impl.py", line 121, in _do_get
    return self._pool.get(wait, self._timeout)
  File "/usr/lib64/python3.7/site-packages/sqlalchemy/util/queue.py", line 162, in get
    self.not_empty.wait(remaining)
  File "/usr/lib64/python3.7/threading.py", line 300, in wait
    gotit = waiter.acquire(True, timeout)
  File "/usr/lib/python3.7/site-packages/eventlet/semaphore.py", line 107, in acquire
    hubs.get_hub().switch()
  File "/usr/lib/python3.7/site-packages/eventlet/hubs/hub.py", line 313, in switch
    return self.greenlet.switch()
```
## Queue
```
    def get(self, block=True, timeout=None):
        self.not_empty.acquire()
        try:
            ...
        finally:
            self.not_empty.release()
```
green thread could be switch in acquire, but not relase when timeout.

**Reply by CaselIT** [reply]
this case is only on the 1.3 branch, it does not affect 1.4 or v2.
You should look into updating to at least the last 1.3 version, or 1.4 if possible

**Reply by Archer-Code** [reply]
Required to fix based on version 1.3 for some reason.  I'm just sharing the mistakes, and it would be nice if someone could answer them.

**Reply by Archer-Code** [reply]
I modified the method of the Queue.get so that it catches the eventlet.timeout exception.  Maybe this will solve the problem I'm having.

**Reply by CaselIT** [reply]
> Required to fix based on version 1.3 for some reason. I'm just sharing the mistakes, and it would be nice if someone could answer them.

not sure what you are asking here. 
by all means do share additional issues you may encounter so we can check if they apply for supported versions of sqlalchemy.
